tools/serve/serve.py

```diff
@@ -52,6 +52,32 @@ def domains_are_distinct(a, b):
     return a_parts[slice_index:] != b_parts[slice_index:]
 
 
+class HtmlScriptInjectorWrapper:
+
+    def __init__(self, base_path=None, url_base="/", inject=""):
+        self.base_path = base_path
+        self.url_base = url_base
+        self.inject = inject
+        self.handler = handlers.handler(self.handle_request)
+
+    def __call__(self, request, response):
+        rv = self.handler(request, response)
+        return rv
+
+    def handle_request(self, request, response):
+        path = filesystem_path(self.base_path, request, self.url_base)
+
+        try:
+            with open(path, 'r') as f:
+                content = f.readlines()
+            # Insert injected script after DOCTYPE if present.
+            insert_index = 1 if len(content) > 0 and content[0].startswith("<!") else 0
+            content.insert(insert_index, "<script src=\"%s\"></script>\n" % (self.inject))
```

Comment:
Hmm, the doctype isn't necessarily a single line. There aren't many cases of it, but here's one:

https://github.com/web-platform-tests/wpt/blob/cad9fcc322e430a7be43e1ef72d685bf516ec46f/html/dom/documents/resource-metadata-management/document-compatmode-02.html#L1-L2

If WebDriver supported it, we'd really want to use a feature like https://github.com/w3c/webdriver-bidi/issues/65 here, but that's not an option right now.

Did you look into using testharness.js or testharnessreport.js as the delivery mechanism instead? We could prepend something to either of those, although inserting parsing-blocking scripts would be a hassle. It comes down to how we'd want to maintain the polyfills, and whether we we need to support reftests which don't import testharness.js.

Reply:
I think it would be really valuable to support reftests as well, as these often represent important real world situations (e.g. does my scroll linked animation look the way it should). The other nice benefit of injecting the polyfill in the page is that it's virtually identical to the real world deployment scenario where the polyfill would also be included early on in the page.

Reply:
In the latest patch I used a regular expression substitution which handles multiline doctypes.